File: c/problem_0017.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include "./lib/mathlib.h"

/**
 * @see https://projecteuler.net:
 *
 * Number letter counts
 * Problem 17
 *
 * If the numbers 1 to 5 are written out in words: one, two, three, four, five, then there are 3 + 3 + 5 + 4 + 4 = 19 letters used in total.
 *
 * If all the numbers from 1 to 1000 (one thousand) inclusive were written out in words, how many letters would be used?
 *
 * NOTE: Do not count spaces or hyphens. For example, 342 (three hundred and forty-two) contains 23 letters and 115 (one hundred and fifteen) contains 20 letters.
 * The use of "and" when writing out numbers is in compliance with British usage.
 */

int main(void)
{
    int single[] = {
        0, // zero, never said
        3, // one
        3, // two
        5, // three
        4, // four
        4, // five
        3, // six
        5, // seven
        5, // eight
        4, // nine
        3, // ten
        6, // eleven
        6, // twelve
        8, // thirteen
        8, // fourteen
        7, // fifteen
        7, // sixteen
        9, // seventeen
        8, // eighteen
        8  // nineteen
    };

    int tenths[] = {
        0, // not counted
        3, // ten, but should not be used
        6, // twenty
        6, // thirty
        5, // forty
        5, // fifty
        5, // sixty
        7, // seventy
        6, // eighty
        6  // ninety
    };

    int hundred = 7; // "hundred"
    int counter = 0, i;

    for (int h = 0; h < 10; h++)
    {
        // h is the hundred counter
        if (h > 0)
        {
            // count e.g. "one hundred"
            counter += single[h] + hundred;
        }

        // count up to (but not including) twenty, as those are special:
        for (i = 1; i < 20; i++)
        {
            if (h > 0)
            {
                // count e.g. "one hundred"
                counter += single[h] + hundred;
                counter += 3; // the "and" from "one hundred and xxx"
            }
            // count e.g. 'twelve"
            counter += single[i];
        }

        for (i = 20; i < 100; i++)
        {
            if (h > 0)
            {
                // count e.g. "one hundred"
                counter += single[h] + hundred;
                counter += 3; // the "and" from "one hundred and xxx"
            }
            // count from 20 to 99:
            counter += tenths[i / 10]; // tenth, e.g. "twenty"
            counter += single[i % 10]; // e.g. "one"
        }
    }

    counter += 11; // "one thousand"

    printf("All words from 1 to 1000 have %d letters.\n", counter);
}

File: haskell/src/Problem0017.hs
module Problem0017 (solution) where

{--------------------------------------------------------------------
<p>If the numbers $1$ to $5$ are written out in words: one, two, three, four, five, then there are $3 + 3 + 5 + 4 + 4 = 19$ letters used in total.</p>
<p>If all the numbers from $1$ to $1000$ (one thousand) inclusive were written out in words, how many letters would be used? </p>
<br><p class="note"><b>NOTE:</b> Do not count spaces or hyphens. For example, $342$ (three hundred and forty-two) contains $23$ letters and $115$ (one hundred and fifteen) contains $20$ letters. The use of "and" when writing out numbers is in compliance with British usage.</p>
---------------------------------------------------------------------}

import Data.Map qualified as Map
import Data.Maybe (fromJust)
import EulerLib (trim)
import EulerLib qualified as E

solution :: E.ProblemSolution
solution =
  E.ProblemSolution
    { E.psNr = 17,
      E.psTitle = "Number Letter Counts",
      E.psSolve = show . calcSolution,
      E.psSolution = "",
      E.psReadInput = readFile "data_0017.txt"
    }

nrToWords :: Map.Map Int String -> Int -> String
nrToWords wMap n
  | n < 1 = ""
  | n < 20 = fromJust (Map.lookup n wMap)
  | n < 100 && n `mod` 10 == 0 = fromJust (Map.lookup n wMap)
  | n < 100 = nrToWords wMap ((n `div` 10) * 10) ++ nrToWords wMap (n `mod` 10)
  | n < 1000 =
      let hundredVal = (n `div` 100)
          rest = n `mod` 100
          hundred = fromJust (Map.lookup 100 wMap)
          and =
            if rest > 0
              then "and"
              else ""
          lowerPart =
            if rest > 0
              then nrToWords wMap rest
              else ""
       in nrToWords wMap hundredVal ++ hundred ++ and ++ lowerPart
  | n < 10000 =
      let thousandVal = (n `div` 1000)
          rest = n `mod` 1000
          thousand = fromJust (Map.lookup 1000 wMap)
          lowerPart =
            if rest > 0
              then nrToWords wMap rest
              else ""
       in nrToWords wMap thousandVal ++ thousand ++ lowerPart
  | otherwise = "Unknown"

calcSolution :: String -> Int
calcSolution input = (sum . map length) nrOfWords
  where
    l = lines input
    pairs = map (\l -> (read (l !! 0) :: Int, trim (l !! 1))) (map (E.splitAtChar ':') l)
    wordMap = Map.fromList pairs
    mappedNrToWords = nrToWords wordMap
    inputList = [1 .. 1000]
    nrOfWords = map mappedNrToWords inputList
